### ISSUE TITLE
Bokeh GeoJSON error can damage a jupyter a notebook

### ISSUE THREAD

**Issue by mangecoeur** [bokeh/bokeh#5464]
When trying to load a 15MB GeoJSON file using `GeoJSONDataSource(geojson=geojson)` I accidentally passed a list instead of a string. This triggered `ValueError expected a value of type str, got ...` which then dumped the WHOLE GeoJSON data into the notebook error output. Notebook file ballooned to 30MB from 400kB and crashed the browser because of the volume of output. Since the browser crashed on open, it was impossible to recover the notebook from within Jupyter. Was only able to recover by opening raw ipynb in SublimeText and deleting output from cell.

Bokeh should take care never to dump whole inputs, especially with GIS data which can be very large.

**Comment by bryevdv** [comment]
Hi @mangecoeur thanks for the report. I would disagree that Bokeh should never do this, in fact I think in the vast majority of cases, providing the values inline is and *enormous* aid to understanding and interpreting error messages. But you are correct that in some very specific cases, such as this one, that it is not appropriate. The oversight is simply due to the fact that none of the core devs are heavy GIS users, which is why we rely on reports like this to help make Bokeh better.

**Comment by bryevdv** [comment]
@mangecoeur can you supply a sample GeoJSON file and code to help with developing this change?

**Comment by mangecoeur** [comment]
@bryevdv  You're right that inputs can help for debug but I suspect with visualisation most inputs are going to be potentially large arrays outside of very simple cases - perhaps Bokeh could estimate the size of inputs and truncate over a threshold (maybe a verbosity control somewhere).

I've attached the GeoJSON in question, created from shapefile  of EU admin regions available free for non-commercial use [here](http://ec.europa.eu/eurostat/web/gisco/geodata/reference-data/administrative-units-statistical-units/nuts#nuts06)


[geojson_nuts.json.zip](https://github.com/bokeh/bokeh/files/607093/geojson_nuts.json.zip)

**Comment by bryevdv** [comment]
@mangecoeur that's exactly the approach I was imagining (and similar to that NumPy, etc.) do as well. Thanks for the file

**Comment by bryevdv** [comment]
OK I tested a tentative implementation the behaves like this:
```
In [3]: r.start = "a"*400000000
```
which results in (with some manual line breaks added be me)
```
ValueError: expected an element of either Float, Datetime or Int, got 
'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa <399999800 characters omitted> aaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
```

Some things to note:

"bad" values can be anything, so the simplest to implement strategy is just to still generate `str(value)` using all the memory that requires, but not print all of it above some threshold. In a perfect world the `str(value)` could possibly be avoided, but I think it is just too much fragile code to try and to that. In any case I just made 400M character string and it only took ~2 seconds. given that it will never overwhelm the notebook, this seems a clear improvement, even if it is not completely perfect. 

I will work up a fuller PR with a `BOKEH_PROPERTY_ERROR_MAX` config env var and tests, etc.

**Comment by mattpap** [comment]
> I will work up a fuller PR with a BOKEH_PROPERTY_ERROR_MAXLEN config env var and tests, etc.

Does this really have to be configurable? Who will be adjusting it?

**Comment by mattpap** [comment]
This seems like a problem with the notebook as well, because you can easily issue a command like `np.array(['a'*10000000])` with similar results. Notebook should be aware that dumping tens or hundreds of megabytes of data into DOM isn't going to work well.

**Comment by mangecoeur** [comment]
@mattpap Property does need to be configurable for debugging etc. Also while it is an issue for the notebook to handle long outputs, there is a  difference between intentionally generating them and having them produced accidentally by libraries. Probably would be interesting to have the notebook protect itself from excessive outputs but at least in this case it's safer is bokeh can avoid adding to the problem.

**Comment by mattpap** [comment]
> Property does need to be configurable for debugging

I doubt making it configurable will help with debugging. If the limit is set to some reasonable value, then on those rare occasions that it's unhelpful, then it will be faster to set up a breakpoint and debug the code, than incrementing the value and waiting for output in a loop. But maybe it's just more efficient to me.

> numpy, pandas already truncate their own outputs in many cases.

This is why I used `np.array(['a'*10000000])`, which doesn't truncate, as an example. Notebook should protect against it as it is really easy to stall or even crash it.

**Comment by bryevdv** [comment]
I would rather be able to just set an env var than going digging for where to put a debugger. Note this applies outside the notebook too, if people are logging, etc they might want to be able to tune the limit to their particulars. I have the start of this already I will try to push a branch later today.

**Comment by bryevdv** [comment]
Another user has run into this: https://discourse.bokeh.org/t/endless-error-message-with-fileinput-and-big-file/9438

> when using FileInput with a server callback, error messages will always display the base64 string . If this string is big (100mb in my case) debugging gets impossible, because it prints out the complete string before any error.
Is there a way to prevent this?

This applies outside the notebook so I don't actually think the notebook is relevant for this issue. However, this issue was created well before `FileInput` and I wonder if `FileInput` in particular might warrant specialized behavior. I would speculate that it is preferable to avoid spilling file contents into logs. I'm not sure a simple maxlen is sufficient to capture the possibilities that are needed to cover. Interested in suggestions @bokeh/dev
